libraries/standard/mqtt/src/mqtt.c

```diff
@@ -75,6 +77,387 @@ static int32_t sendPacket( MQTTContext_t * pContext,
+ *
+ * @param[in] packetType First byte of fixed header.
+ *
+ * @return Type of ack.
+ */
+static MQTTPubAckType_t getAckFromPacketType( uint8_t packetType )
+{
+    MQTTPubAckType_t ackType;
+    switch( packetType )
+    {
+        case MQTT_PACKET_TYPE_PUBACK:
+            ackType = MQTTPuback;
+            break;
+        case MQTT_PACKET_TYPE_PUBREC:
+            ackType = MQTTPubrec;
+            break;
+        case MQTT_PACKET_TYPE_PUBREL:
+            ackType = MQTTPubrel;
+            break;
+        case MQTT_PACKET_TYPE_PUBCOMP:
+            ackType = MQTTPubcomp;
+            break;
+        default:
+            /* Not an ack. */
+            assert( 0 );
+            break;
+    }
+    return ackType;
+}
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Dump a packet from the transport interface.
+ *
+ * @param[in] PContext MQTT Connection context.
+ * @param[in] remainingLength Remaining length of the packet to dump.
+ * @param[in] bytesAlreadyReceived Bytes previously read from the transport.
+ *
+ * @return #MQTTRecvFailed or #MQTTNoDataAvailable.
+ */
+static MQTTStatus_t dumpPacket( MQTTContext_t * const pContext,
+                                size_t remainingLength,
+                                int32_t bytesAlreadyReceived )
+{
+    MQTTStatus_t status = MQTTRecvFailed;
+    int32_t totalBytesReceived = bytesAlreadyReceived, bytesReceived = 0;
+    assert( pContext != NULL );
+    size_t bytesToReceive = pContext->networkBuffer.size;
+
+    while( totalBytesReceived < ( int32_t ) remainingLength )
+    {
+        /* Update number of bytes to receive. */
+        if( ( remainingLength - ( size_t ) totalBytesReceived ) < bytesToReceive )
+        {
+            bytesToReceive = remainingLength - ( size_t ) totalBytesReceived;
+        }
+
+        bytesReceived = pContext->transportInterface.recv( pContext->transportInterface.networkContext,
+                                                           pContext->networkBuffer.pBuffer,
+                                                           bytesToReceive );
+        if( bytesReceived != ( int32_t ) bytesToReceive )
+        {
+            /* Partial receive occurred while trying to dump packet. */
+            IotLogErrorWithArgs( "Partial receive while dumping packet."
+                                 "Received bytes=%d, Expected bytes=%u",
+                                 bytesReceived,
+                                 bytesToReceive );
+            status = MQTTRecvFailed;
+            break;
+        }
+
+        /* Update number of bytes to receive. */
+        totalBytesReceived += bytesReceived;
+    }
+
+    if( totalBytesReceived == ( int32_t ) remainingLength )
+    {
+        IotLogInfoWithArgs( "Dumped packet. Dumped bytes=%d",
+                            totalBytesReceived );
+        /* Packet dumped, so no data is available. */
+        status = MQTTNoDataAvailable;
```

Comment:
This status code indicate that packet could not fit the provided buffer, so that application can take some action like increase buffer size.  MQTTNoDataAvailable does not really convey that message.

Reply:
The intention here was to log a warning when the packet exceeds the buffer, dump the packet, and then continue the process loop. That's why the status is reset to `MQTTNoDataAvailable`, to indicate that there is no data left to be read and to continue the process loop should continue. However, if the intention is to stop the loop after dumping a packet, then I agree in changing this status, as it would give an error code back to the application.

Reply:
Changed to `MQTTBufferExceeded` so error will bubble back to application.

Reply:
OK I looked up V3 as well as v1.4x. and the code silently drops the packet if it exceeds the buffer ( no bubble up) . (https://github.com/aws/aws-iot-device-sdk-embedded-C/blob/f50ae301378ad2546b966521d433ebfbc185c4bf/src/aws_iot_mqtt_client_common_internal.c#L450) After giving some thought, I like the earlier approach. We should change it back to the original behavior and log error, so that  we don't terminate the connection on spurious large packets.

Reply:
Reverted